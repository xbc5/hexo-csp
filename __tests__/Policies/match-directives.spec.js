"use strict";

const { initial, factory, arg } = require("./util");
const clone = require("lodash.clonedeep");

["match", "directives"].forEach((method) => {
  describe(`${method}()`, () => {
    it("should be defined", async () => {
      const result = factory()[method];
      expect(result).toBeDefined();
    });
  });
});

[
  {
    arg: undefined,
    initial: initial(),
    method: "match",
    strategy: "the constructor",
  },
  { arg: arg(), initial: undefined, method: "match", strategy: "a method arg" },
  {
    arg: undefined,
    initial: initial(),
    method: "directives",
    strategy: "the constructor",
  },
  {
    arg: arg(),
    initial: undefined,
    method: "directives",
    strategy: "a method arg",
  },
].forEach(({ arg, initial, strategy, method }) => {
  describe(`${method}()`, () => {
    describe(`when passing policies via ${strategy}`, () => {
      it("should return the matched policy", async () => {
        const result = factory({ initial })[method]("foo", arg);
        expect(result).toMatchSnapshot();
      });

      it("should return multiple matched policies", async () => {
        const reset = (target) => {
          const r = clone(target);
          r[0].pattern = "foo";
          r[1].pattern = "bar";
          return r;
        };

        let i, a;
        if (initial) i = reset(initial);
        if (arg) a = reset(arg);

        const result = factory({ initial: i })[method]("foobar", a);
        expect(result).toMatchSnapshot();
      });

      it("should not return any policies if not matched", async () => {
        const result = factory({ initial })[method]("xyz", arg);
        expect(result).toMatchSnapshot();
      });
    });
  });
});
