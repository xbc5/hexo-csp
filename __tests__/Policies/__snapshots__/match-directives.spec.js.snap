// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`directives() when passing policies via a method arg should not return any policies if not matched 1`] = `Object {}`;

exports[`directives() when passing policies via a method arg should return multiple matched policies 1`] = `
Object {
  "default-src": Array [
    "https://default-foo-arg.com",
    "https://default-bar-arg.com",
  ],
  "img-src": Array [
    "https://img-foo-arg.com",
    "https://img-bar-arg.com",
  ],
}
`;

exports[`directives() when passing policies via a method arg should return the matched policy 1`] = `
Object {
  "default-src": Array [
    "https://default-foo-arg.com",
  ],
  "img-src": Array [
    "https://img-foo-arg.com",
  ],
}
`;

exports[`directives() when passing policies via the constructor should not return any policies if not matched 1`] = `Object {}`;

exports[`directives() when passing policies via the constructor should return multiple matched policies 1`] = `
Object {
  "default-src": Array [
    "https://default-foo-base.com",
    "https://default-bar-base.com",
  ],
  "img-src": Array [
    "https://img-foo-base.com",
    "https://img-bar-base.com",
  ],
}
`;

exports[`directives() when passing policies via the constructor should return the matched policy 1`] = `
Object {
  "default-src": Array [
    "https://default-foo-base.com",
  ],
  "img-src": Array [
    "https://img-foo-base.com",
  ],
}
`;

exports[`match() when passing policies via a method arg should not return any policies if not matched 1`] = `Array []`;

exports[`match() when passing policies via a method arg should return multiple matched policies 1`] = `
Array [
  Object {
    "directives": Object {
      "default-src": Array [
        "https://default-foo-arg.com",
      ],
      "img-src": Array [
        "https://img-foo-arg.com",
      ],
    },
    "mode": "merge",
    "pattern": "foo",
  },
  Object {
    "directives": Object {
      "default-src": Array [
        "https://default-bar-arg.com",
      ],
      "img-src": Array [
        "https://img-bar-arg.com",
      ],
    },
    "mode": "merge",
    "pattern": "bar",
  },
]
`;

exports[`match() when passing policies via a method arg should return the matched policy 1`] = `
Array [
  Object {
    "directives": Object {
      "default-src": Array [
        "https://default-foo-arg.com",
      ],
      "img-src": Array [
        "https://img-foo-arg.com",
      ],
    },
    "mode": "merge",
    "pattern": "^foo$",
  },
]
`;

exports[`match() when passing policies via the constructor should not return any policies if not matched 1`] = `Array []`;

exports[`match() when passing policies via the constructor should return multiple matched policies 1`] = `
Array [
  Object {
    "directives": Object {
      "default-src": Array [
        "https://default-foo-base.com",
      ],
      "img-src": Array [
        "https://img-foo-base.com",
      ],
    },
    "mode": "merge",
    "pattern": "foo",
  },
  Object {
    "directives": Object {
      "default-src": Array [
        "https://default-bar-base.com",
      ],
      "img-src": Array [
        "https://img-bar-base.com",
      ],
    },
    "mode": "merge",
    "pattern": "bar",
  },
]
`;

exports[`match() when passing policies via the constructor should return the matched policy 1`] = `
Array [
  Object {
    "directives": Object {
      "default-src": Array [
        "https://default-foo-base.com",
      ],
      "img-src": Array [
        "https://img-foo-base.com",
      ],
    },
    "mode": "merge",
    "pattern": "^foo$",
  },
]
`;
